0.9.292 (Admin roles)

redsummernight opened the pull request:
Ultimate release review.

Commit:
AO3-5518 Create roles for admins (#3821)

* AO3-5519: Add roles to admins and restrict access to news posting

* Fix div id for admin header links

* Fine-tuning admin error messages and fixing test failures

* Remove unused admins controller and routes

* AO3-5519: Extend admin roles to the settings page, use pundit a little more organically, centralize admin methods, and tweak some tests

* Update some invitation tests

* AO3-5519: Add auth to admin user and content areas, fix small bug in link, remove unused controller action

* AO3-5519: Add new admin landing page, remove redundant menu, and adjust tests

* AO3-5519: More test whackamole

* Checking something in Travis

* Checking something else in Travis

* Third time's the charm

* Attempting a louder failure

* Update admin factory syntax

* Adds message to alert admins they have no roles

* changes name of abuse role in admin policy

* Updates schema with latest migrations

* Adds admin roles tests

* code climate syntax fixes

* small syntax change to retrigger tests

* Requested changes from code review

* Reverts schema file

* Review tweaks

* Review suggestions

* Removes empty parans after admin only method

* Review request change, modifies call to policy in admin header

* Removes undefined step in admin step definition

* Updates admin moderation policy

* Review changes requests

Co-authored-by: elz <elzj78@gmail.com>
Co-authored-by: sarken <sarken@gmail.com>
Co-authored-by: Matt Sears <mattsears@users.noreply.github.com>

sarken committed:
AO3-5518 Tweak admin index HTML and add i18n (#3855)

* AO3-5518 Add missing class names and divs to admin index

* AO3-5518 i18n the admins index

* AO3-5518 i18n role names

tommilligan committed:
AO3-5979 Add unique constraints for login and email on the admins table (#3856)

* AO3-5979 remove duplicate admin email validation

* AO3-5979 add unique indices to admin login and email

Commit:
AO3-5519 AO3-5883 AO3-5886 Restrict admin access to admin posts, settings, and comments (#3836)

* AO3-5519: Add roles to admins and restrict access to news posting

* Fix div id for admin header links

* Fine-tuning admin error messages and fixing test failures

* Remove unused admins controller and routes

* AO3-5519: Extend admin roles to the settings page, use pundit a little more organically, centralize admin methods, and tweak some tests

* Update some invitation tests

* AO3-5519: Add auth to admin user and content areas, fix small bug in link, remove unused controller action

* AO3-5519: Add new admin landing page, remove redundant menu, and adjust tests

* AO3-5519: More test whackamole

* Checking something in Travis

* Checking something else in Travis

* Third time's the charm

* Attempting a louder failure

* Update admin factory syntax

* Adds message to alert admins they have no roles

* changes name of abuse role in admin policy

* Updates schema with latest migrations

* Adds admin roles tests

* code climate syntax fixes

* small syntax change to retrigger tests

* Requested changes from code review

* Reverts schema file

* Review tweaks

* Review suggestions

* Removes empty parans after admin only method

* AO3-5519 Adds tests for admin functionality

* AO3-5883 Removes admin buttons for admins without permissons from external works and bookmakrs

AO3-5883 Adds tests for user creations controller

AO3-5883 Modifying permissions around admin actions

AO3-5883 Makes pundit_user availble to all views, to allow use in required contexts

AO3-5883 modifies bookmarking step

* AO3-5883 Updates settings page to fit requirements

AO3-5883 Adds tests for settings controller

AO3-5883 changes the way errors are rendered in the setting update method

AO3-5883 ixes settings controller tests

* AO3-5886 Builds policy to restict comment actions

AO3-5886 Adds tests for admin functionality

AO3-5886 Adds tests, modifies logic for comment policies

* AO3-5886 Uses pundit policy method instead of admin model method

AO3-5886 Adds test for comments controller

* Review request change, modifies call to policy in admin header

* Review changes/ code quality fixes

* Removes undefined step in admin step definition

* Removes undefined step in admin step definition

* refactors methods in admin setting policy

* code formatting refactor

* another refactor of admin setting policy

* Review change requests

* Updates admin moderation policy

* Retrigger tests

* Review changes requests

* Review changes

* Review suggestions

* AO3-5886 Only certain admins should be able to mark comments as ham (never users)

* AO3-5519 AO3-5883 AO3-5886 Update spec descriptions

* Resolving code climate issue

* Reverts settings controller check

* Removes normal user references from comment policy

* AO3-5886 Remove unnecessary ownership check for marking comments as spam

Author checks are handled by the before actions.

Also, remove the can_mark_comment_ham? method in comment_policy, and
update helper method for displaying Spam and Not Spam buttons to
reflect that the can_mark_comment_ham? and can_mark_comment_spam?
methods are both removed.

* AO3-5518 Make it clear ApplicationPolicy is only for admins

* AO3-5886 Address Hound complaint about missing blank line in comment controller spec

* fixes hound issues

Co-authored-by: elz <elzj78@gmail.com>
Co-authored-by: sarken <sarken@gmail.com>
Co-authored-by: Matt Sears <mattsears@users.noreply.github.com>

Commit:
AO3-5884 Restrict ability to search for and update users (#3838)

* AO3-5519: Add roles to admins and restrict access to news posting

* Fix div id for admin header links

* Fine-tuning admin error messages and fixing test failures

* Remove unused admins controller and routes

* AO3-5519: Extend admin roles to the settings page, use pundit a little more organically, centralize admin methods, and tweak some tests

* Update some invitation tests

* AO3-5519: Add auth to admin user and content areas, fix small bug in link, remove unused controller action

* AO3-5519: Add new admin landing page, remove redundant menu, and adjust tests

* AO3-5519: More test whackamole

* Checking something in Travis

* Checking something else in Travis

* Third time's the charm

* Attempting a louder failure

* Update admin factory syntax

* Adds message to alert admins they have no roles

* changes name of abuse role in admin policy

* Updates schema with latest migrations

* Adds admin roles tests

* code climate syntax fixes

* small syntax change to retrigger tests

* Requested changes from code review

* Reverts schema file

* Review tweaks

* Review suggestions

* Removes empty parans after admin only method

* AO3-5884 Modifies logic around user management to fit requirements

AO3-5884 adding tests for admin user management

AO3-5884 Adds tests for admin user actions

AO3-5884 Fixes params breaking tests

AO3-5884 Fixing tests for admin user management features

* AO3-5884 Adds invitation test

AO3-5884 Adds invitation tests for all controller methods

AO3-5884 Adding tests for features

AO3-5884 Adds tests for invitations find method

* Review request change, modifies call to policy in admin header

* Removes undefined step in admin step definition

* Addresses some code formatting issues in user_manager spec

* cleans up code formatting in tests

* Code formatting fixes in tests

* fixing formatting issues in tests

* Rerunning test

* Updates admin moderation policy

* updates to admin moderation policy

* Review changes requests

* Formatting fixes in tests

* switching single quotes for doubles

* AO3-5884 Update specs

* AO3-5884 Update user_manage_spec contexts

* Review requests

Co-authored-by: elz <elzj78@gmail.com>
Co-authored-by: sarken <sarken@gmail.com>
Co-authored-by: Matt Sears <mattsears@users.noreply.github.com>

Commit:
AO3-5885 Restrict ability modify user creations (#3839)

* AO3-5519: Add roles to admins and restrict access to news posting

* Fix div id for admin header links

* Fine-tuning admin error messages and fixing test failures

* Remove unused admins controller and routes

* AO3-5519: Extend admin roles to the settings page, use pundit a little more organically, centralize admin methods, and tweak some tests

* Update some invitation tests

* AO3-5519: Add auth to admin user and content areas, fix small bug in link, remove unused controller action

* AO3-5519: Add new admin landing page, remove redundant menu, and adjust tests

* AO3-5519: More test whackamole

* Checking something in Travis

* Checking something else in Travis

* Third time's the charm

* Attempting a louder failure

* Update admin factory syntax

* Adds message to alert admins they have no roles

* changes name of abuse role in admin policy

* Updates schema with latest migrations

* Adds admin roles tests

* code climate syntax fixes

* small syntax change to retrigger tests

* Requested changes from code review

* Reverts schema file

* Review tweaks

* Review suggestions

* Removes empty parans after admin only method

* AO3-5885 Removes admin buttons for admins without permissons from external works and bookmarks

AO3-5885 adds tests for user creations controller

AO3-5885 Modifies permissions around admin actions

* AO3-5885 Makes pundit_user availble to all views, to allow use in required contexts

* AO3-5885 modifies bookmarking step

* AO3-5885 Adds tests for spam controller

* Review request change, modifies call to policy in admin header

* Removes undefined step in admin step definition

* Code formatting fix

* Updates admin moderation policy

* updates admin moderation policy

* Review changes requests

* Switch out single quotes for doubles

* Formatting tweaks in spam controller test

* Review suggestions

* Fixing tests

* AO3-5885 Spec updates

Expand coverage a smidge, update descriptions, make it a bit easier to
add tests for other types of creations in the future by adding contexts

* Fixes hound issues in spam tests

Co-authored-by: elz <elzj78@gmail.com>
Co-authored-by: sarken <sarken@gmail.com>
Co-authored-by: Matt Sears <mattsears@users.noreply.github.com>

redsummernight committed:
AO3-5980 AO3-5623 Update admin creation script to prompt for roles (#3857)

* AO3-5980 AO3-5623 Update admin creation script to prompt for roles

Also:

- Print the admin login HTTPS URL for the right environment
- Remove spaces in admin logins
- Output i18n role names in wiki output

* AO3-5980 Allow admin script to update existing admins

- Leave email/roles blank if we don't want to update them for existing admins
- Set password and print password file only for new admins
- Print admins' created times in the wiki output instead of today's date

[skip codeship tests]

sarken committed:
AO3-5885 Rework restrictions on modifying user creations + grant support admins edit access (#3858)

* AO3-5885 Rename AdminModerationPolicy to UserCreationPolicy to reflect what the policy controls permissions for

Also separate by action

* AO3-5885 Use a helper to determine whether to show admin options

Remove some unnecessary HTML and update some Ruby syntax too.

* AO3-5885 Update paths in admin_options partial

* AO3-5885 Use helper for determining whether to show admin hide options

* AO3-5885 Use helper for determining whether to show admin edit options on creation

* AO3-5885 Use helper when determining whether to show admin spam options on user creations

* AO3-5885 Use helper when determining whether to show admin delete option on user creations

* AO3-5885 Allow admins with support role to edit user creations

* AO3-5885 i18n admin options partial

* AO3-5885 Fix code comment typo noted in review

tickinginstant committed:
AO3-5992 Fix error with spam alert email. (#3859)

sarken commented on app/models/user_manager.rb:
```diff
@@ -44,6 +46,17 @@ def error_message
+    if admin_access
+      true
+    else
+      errors << "Must have authorized admin role to proceed"
```
Now that I've seen this in action, I am not a fan. I would like it to be more like the message in https://github.com/redsummernight/otwarchive/pull/6/files#diff-55c5b7aecfb519d0e4880eaf2788eb6eR287, e.g. "Sorry, only an authorized admin can do that."

redsummernight commented on app/views/admin/_header.html.erb:
```diff
@@ -19,15 +22,21 @@
   <li><%= link_to ts("Spam", key: "header"), admin_spam_index_path %></li>
-  <li><%= link_to ts("Settings", key: "header"), admin_settings_path %></li>
+
+  <% if AdminSettingPolicy.can_update_settings?(current_admin) %>
```
We can use `policy(AdminSetting).can_update_settings?` here.

redsummernight commented on app/views/admin/settings/index.html.erb:
```diff
@@ -5,77 +5,87 @@
-    </dl>
-  </fieldset>
-  
+  <% if AdminSettingPolicy.user_has_correct_role?(current_admin, nil) %>
```
Admins with any of the settings roles (superadmin, support, policy_and_abuse, or tag_wrangling) should be able to see all settings, even if they can update some of them.

We can disable the inputs of the settings they cannot update, instead of hiding them entirely.

redsummernight commented on app/policies/admin_user_policy.rb:
```diff
@@ -0,0 +1,36 @@
+class AdminUserPolicy < ApplicationPolicy
```
This looks unused right now(?).

sarken replied:
Yeah, I think UserPolicy is the one we actually want to keep. This is like... a previous version of that.

redsummernight committed:
AO3-5883 Show restricted admin settings as disabled fields (#3860)

* AO3-5883 Fix Style/Alias

* AO3-5883 Update settings of Style/PercentLiteralDelimiters

* AO3-5883 Use Pundit's strong parameters handling

- Show all settings on the form to any admin who can edit some of the settings.
- Remove list of settings from permission error message (simpler i18n later).
- Disallow parameter last_updated_by from controller.
- Use AdminSetting.default instead of adding a factory.

* AO3-5883 Disable settings fields the current admin cannot update

* AO3-5883 i18n admin settings

* AO3-5883 Remove permission error, clean up settings form

sarken committed:
AO3-5886 Separate permissions in CommentPolicy and add notice to comments marked as spam (#3861)

* AO3-5886 Have separate roles for deleting comments and updating their spam status

Rename 'delete' to 'destroy' to match UserCreationPolicy

* AO3-5886 Move logic for displaying comment delete button into a helper

* AO3-5886 Add notice indicating a comment has been marked as spam

This is necessary because all admins, regardless of role, can see spam comments. However, if
they do not have a role that lets them see the Not Spam button, they have no idea if a comment
has been marked as spam or not.

* AO3-5886 Remove unused methods from CommentPolicy

* AO3-5886 Fix error in resolving merge conflict

redsummernight committed:
AO3-5884 Remove unused class AdminUserPolicy (#3864)

* AO3-5884 Remove unused class AdminUserPolicy

- Omit specifying UserPolicy when authorizing User objects.
- Disable "Abuse Administration" form for admins without access.
- Authorize before the "update_status" action, which means we don't need to
  check for valid admin roles inside UserManager, and the authorization error
  message is the same as everywhere else.
- Update UserManager specs to stop checking admin roles.

Extra:

- Authorize the external work, not the admin, when using UserCreationPolicy.
- Stop passing current_admin as an argument, call the helper directly.
- Remove unused static methods in policy classes.
- Update Rubocop rules RSpec/DescribeClass and Style/PercentLiteralDelimiters.

* AO3-5884 Rename can_troubleshoot_users to can_manage_users

sarken committed:
AO3-5518 AO3-5519 AO3-5886 Add board admin role with admin post and comment moderation permissions (#3865)

* AO3-5518 Add board admin role

Board wants permissions to create admin posts, delete comments, and mark comments spam.

* AO3-5518 Add name of board role to locale file

redsummernight committed:
AO3-5623 AO3-5518 Fix admin creation script, clean up admin-related feature steps (#3866)

* AO3-5980 Fix minor issues with admin creation script

- Remove stray quote.
- Print filename so we can identify error messages.
- Sort wiki output by admins' created timestamps.

* AO3-5518 Clean up feature steps to create admins

- Have one step for all types of admins.
- Use the admin role in the admin's username so we can reuse the admin later
  when we need the same role.
- Use a more specific role than "superadmin" whenever we can.
- Use the generic no-role admin if the feature under test is not yet locked
  behind any admin roles.
- Remove tests where admins reply to news post comments, as we won't do AO3-3685.

* AO3-5518 Remove unused assignment

* AO3-5518 Use admins without roles if possible

zz9pzza committed:
AO3-5967 Ignore http_proxy for requests to New Relic collectors (#3849)

From the New Relic Ruby agent development team:
https://discuss.newrelic.com/t/avoiding-a-global-proxy-config/104464/4

tommilligan committed:
AO3-5793 Docker support (#3863)

* AO3-5793 Add Dockerfile, init scripts and config

* AO3-5793 Update existing config to work with docker

* AO3-5793 Update bookmarks controller

* feedback, nits from review

* standardise and improve loading memcached servers

* docker: move files into script and config subdirs, update scripts

* add docstring to docker-compose

sarken committed:
AO3-5884 Restrict which admins can update user roles and which can update user emails (#3869)

* AO3-5884 Restrict which user attributes admins can update

Tag Wrangling and Open Doors need to update user roles, but not emails.
Support and Policy & Abuse need to update user emails, but not roles.

* AO3-5884 Add additional code comment

* AO3-5884 Disable user role and email field when admin doesn't have permission to update them

* AO3-5884 Allow removal of all roles

* AO3-5884 Use empty string for value of hidden roles field

* AO3-5884 Prefer include matcher in specs

* AO3-5884 Style fixes to spec

* AO3-5884 More style fixes to admin_users_controller_spec

* AO3-5884 Update tests that appeared after merging with master

* AO3-5884 Prefer [] to ()

tickinginstant committed:
AO3-5999 Make comment factories look more like real comments, and reorganize to use traits. (#3870)

* AO3-5999 Fix up comment factories.

* AO3-5999 Make more steps use factories.

* AO3-5999 Fix the chaptered work step.

* AO3-5999 Generate tag comments with factories.

* AO3-5999 Fix chapter positions in work step.

sarken committed:
AO3-5883 Let policy_and_abuse admins turn queue on/off and specify how many invites to send per day (#3871)

* AO3-5883 Allow policy_and_abuse admins to update some invitation-related settings

These are the only two settings they recall using, but they can have more if they want.

* AO3-5883 Remove extra comma

* AO3-5883 Don't use default number of invites per day in spec that checks that setting has been updated

* AO3-5883 Prefer policy_and_abuse admin to superadmin in tests

* AO3-5883 Prefer field labels in cukes that involve filling out forms

* AO3-5883 ...add file I forgot

redsummernight committed:
AO3-5983 Add Hound rules for Rails and ERB template linting (#3868)

* AO3-5983 Add .erb-lint.yml

* AO3-5983 Update rubocop to 0.83

* AO3-5983 Add rubocop-rails

* AO3-5983 Tweaks

* AO3-5983 Lock linter versions

* AO3-5983 Remove addendum for DeprecatedClasses

[skip codeship tests]

* AO3-5983 Prefer square brackets for %w delimiters

* AO3-5983 Remove version for erb_lint

Hound only supports one version (see their docs), we shouldn't specify any.

* AO3-5983 Fix style violations in the latest release

redsummernight committed:
AO3-5884 Fix deleting roles when admins can't update roles (#3873)

* AO3-5884 Add failed tests for PAC admins not changing roles

* AO3-5884 Fix deleting roles when admins don't have permissions to update roles

Previously, if an admin does not have permissions to update roles, and we don't
include "roles" in the permitted attributes, the user's roles will be set to
an empty array. Now we skip setting roles entirely in that case.

Also stop skipping validations when updating users, and show validation errors
to admins, so admins can't accidentally break users by saving blank emails, etc.

* AO3-5884 Rename matcher not_change to avoid_changing

redsummernight commented:
Hound is stalling, but we can expect some comments about RSpec/InstanceVariable which we decided to ignore in `https://github.com/otwcode/otwarchive/pull/3870`.
